packages/circuits/test/proveNegativeReputation.test.ts

```diff
@@ -0,0 +1,276 @@
+    GLOBAL_STATE_TREE_DEPTH,
+    EPOCH_TREE_DEPTH,
+} from '@unirep/circuits'
+import { Reputation } from '../src/Reputation'
+
+const circuitName = 'proveNegativeReputation'
+const circuitPath = path.join(
+    __dirname,
+    `../circuits/test/${circuitName}_test.circom`
+)
+
+const genCircuitInput = (
+    id: crypto.ZkIdentity,
+    epoch: number,
+    reputationRecords,
+    attesterId,
+    _maxRep?
+) => {
+    const maxRep = _maxRep ?? 0
+    if (reputationRecords[attesterId] === undefined) {
+        reputationRecords[attesterId] = new Reputation(
+            BigInt(0),
+            BigInt(0),
+            BigInt(0),
+            BigInt(0)
+        )
+    }
+
+    // User state tree
+    const userStateTree = new crypto.SparseMerkleTree(
+        USER_STATE_TREE_DEPTH,
+        crypto.hash5([BigInt(0), BigInt(0), BigInt(0), BigInt(0), BigInt(0)])
+    )
+    for (const attester of Object.keys(reputationRecords)) {
+        userStateTree.update(
+            BigInt(attester),
+            reputationRecords[attester].hash()
+        )
+    }
+    const userStateRoot = userStateTree.root
+    const USTPathElements = userStateTree.createProof(BigInt(attesterId))
+
+    // Global state tree
+    const GSTree = new crypto.IncrementalMerkleTree(GLOBAL_STATE_TREE_DEPTH)
+    const commitment = id.genIdentityCommitment()
+    const hashedLeaf = crypto.hashLeftRight(commitment, userStateRoot)
+    GSTree.insert(hashedLeaf)
+    const GSTreeProof = GSTree.createProof(0) // if there is only one GST leaf, the index is 0
+    const GSTreeRoot = GSTree.root
+
+    const circuitInputs = {
+        epoch,
+        identity_nullifier: id.identityNullifier,
+        identity_trapdoor: id.trapdoor,
+        user_tree_root: userStateRoot,
+        GST_path_index: GSTreeProof.pathIndices,
+        GST_path_elements: GSTreeProof.siblings,
+        attester_id: attesterId,
+        pos_rep: reputationRecords[attesterId].posRep,
+        neg_rep: reputationRecords[attesterId].negRep,
+        graffiti: reputationRecords[attesterId].graffiti,
+        sign_up: reputationRecords[attesterId].signUp,
+        UST_path_elements: USTPathElements,
+        maxRep,
+    }
+    return crypto.stringifyBigInts(circuitInputs)
+}
+
+const genProofAndPublicSignals = async (inputs: any): Promise<any> => {
+    const circuitWasmPath = path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.wasm`
+    )
+    const zkeyPath = path.join(__dirname, `../zksnarkBuild/${circuitName}.zkey`)
+    const { proof, publicSignals } = await snarkjs.groth16.fullProve(
+        inputs,
+        circuitWasmPath,
+        zkeyPath
+    )
+    return { proof, publicSignals }
+}
+
+const verifyProof = async (publicSignals, proof): Promise<boolean> => {
+    const vkey = require(path.join(
+        __dirname,
+        `../zksnarkBuild/${circuitName}.vkey.json`
+    ))
+    return snarkjs.groth16.verify(vkey, publicSignals, proof)
+}
+
+const genEpochKey = (
```

Comment:
it can be imported from `@unirep/core`

Reply:
d9b016e30202f701fb47e0e7db4efdfa0e79bb9f